README.md
# vardist
[![forthebadge](http://forthebadge.com/images/badges/as-seen-on-tv.svg)](http://forthebadge.com)
===========

Variable distribution summaries in an easy-to-consume format, and tools to compare variable distributions between data sets.

You're welcome.

## Installation
```r
if (!require("devtools")) { install.packages("devtools") }
devtools::install_github("avantcredit/vardist")
```

## Quick Start

Let's suppose you have a data frame with all numeric fields. If you want variable summaries for this data frame, you can call `calculate_numeric_summaries`:

```r
library(vardist)
df <- data.frame(a = runif(100) + 5, b = runif(100) + 5)
calculate_numeric_summaries(df)
#   variables     mean standard_deviation     min first_quartile   median third_quartile      max percentile_10 percentile_90 missing_pct zero_pct
# 1         a 5.441890          0.2888938 5.01808       5.192760 5.412176       5.691090 5.968628      5.066621      5.893265           0        0
# 2         b 5.505793          0.2887159 5.01823       5.264106 5.542553       5.731539 5.994678      5.127110      5.908680           0        0
```

If you want to compare it to a data frame with baseline values, you can call `compare_numeric_distributions`:
```r
library(vardist)
baseline   <- data.frame(a = runif(100) + 5, b = runif(100) + 5)
challenger <- data.frame(a = runif(100) + 5, b = runif(100) + 8)
compare_numeric_distributions(challenger, baseline)
# $challenger_stats
#   variables     mean standard_deviation      min first_quartile   median third_quartile      max percentile_10 percentile_90 missing_pct zero_pct
# 1         a 5.490797          0.2876895 5.005764       5.254185 5.462860       5.755028 5.994012      5.107292      5.889627           0        0
# 2         b 8.514025          0.2765906 8.000728       8.272374 8.500748       8.765646 8.996240      8.158079      8.889809           0        0
#
# $baseline_stats
#   variables     mean standard_deviation      min first_quartile   median third_quartile      max percentile_10 percentile_90 missing_pct zero_pct
# 1         a 5.508399          0.2935941 5.026837       5.241170 5.469960       5.757797 5.999796      5.123201      5.929961           0        0
# 2         b 5.504826          0.2819114 5.005868       5.254398 5.482953       5.736300 5.972390      5.138694      5.914151           0        0
#
# $report
#   variables  mean standard_deviation  min first_quartile median third_quartile   max percentile_10 percentile_90 missing_pct zero_pct ks_test
# 1         a  TRUE               TRUE TRUE           TRUE   TRUE           TRUE  TRUE          TRUE          TRUE        TRUE     TRUE    TRUE
# 2         b FALSE               TRUE TRUE          FALSE  FALSE          FALSE FALSE         FALSE         FALSE        TRUE     TRUE   FALSE
```

There are analogous functions `calculate_character_summaries` and `compare_character_distributions`, which operate on character fields.


## Custom Summaries and Tests

The package comes with default choices of variable summaries and tests for comparison with the baseline. However, you can write your own custom summaries and tests.

For example, suppose you wanted one of your summaries to be the first element in a column, and the other to be a random element. (In practice, of course, you'd choose *useful* summaries.) Then, for your summaries, you would want:
```r
my_summaries <- list(
  first_element  = function(col) { col[1] },
  random_element = function(col) { col[sample(length(col), 1)] }
)
```

Now suppose that you want to test if the first elements of your data frames are identical, the random elements chosen are identical, and that columns have the same length. (Again, in practice, you'd run *useful* tests.) For your tests, you would want:
```r
my_tests <- list(
  first_element  = function(baseline_value, challenger_value, ...) { identical(baseline_value, challenger_value) },
  random_element = function(baseline_value, challenger_value, ...) { identical(baseline_value, challenger_value) },
  same_col_length = function(baseline_col, challenger_col) { length(baseline_col) == length(challenger_col) }
)
```

The first two elements are named after summaries, so they will have access to the summary values when they are exectued. They will also have access to the underlying columns of data - we write `...` here because we are not using them, but they will still be passed in to the test function.

The third element is not named after any summary, so it only has access to the data itself.

Putting it all together:
```r
library(vardist)
baseline   <- data.frame(a = runif(100) + 5, b = runif(100) + 5)
challenger <- data.frame(a = runif(100) + 5, b = runif(100) + 8)
compare_numeric_distributions(challenger, baseline, summaries = my_summaries, tests = my_tests)
# $challenger_stats
#   variables first_element random_element
# 1         a      5.993870       5.017843
# 2         b      8.494429       8.109733
#
# $baseline_stats
#   variables first_element random_element
# 1         a      5.828025       5.927089
# 2         b      5.727900       5.909929
#
# $report
#   variables first_element random_element same_col_length
# 1         a         FALSE          FALSE            TRUE
# 2         b         FALSE          FALSE            TRUE
```

## License

Copyright 2014-2017 Avant

Author: Abel Castillo

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


NEWS.md
## Version 0.4.1-0.4.2
- Parallelize calculating stats over the columns of the dataframe, as that would
  give a bigger speedup, especially for summaries that take long to calculate.

## Version 0.4.0
- Add support for parallelization via `parallel::mclapply`.

## Version 0.3.4
- Make `within_tolerance` a bit more robust with respect to special cases (handling zeroes and NAs).
- Minor bugfixes between 0.3.0 - 0.3.3

## Version 0.3.0
- renamed `calculate_summaries` and `compare_distributions` to `calculate_numeric_summaries` and `compare_numeric_distributions`.
- Added support for character data frames via `calculate_character_summaries` and `compare_character_distributions`. Use the options "vardist.character_discrete_tolerance", "vardist.character_percent_tolerance", "vardist.character_uniques_threshold", and "vardist.character_uniques_tolerance" to control the default tests.

## Version 0.2.5
- Added options "vardist.summary_tolerance" and "vardist.ks_test_threshold", for use with `default_tests`.

## Version 0.2.4
- Added function `generate_report_from_summaries`.

## Version 0.2.3
- Moved `magrittr` from `Imports` to `Depends`.

## Version 0.2.2
- Enforcing numeric columns only in `calculate_summaries`.

## Version 0.2.1
- Fixed Roxygen documentation, actually exporting functions.

## Version 0.2.0
- Constructed default values for `summaries` and `tests`.
- Added documentation via Roxygen, as well as a more detailed README.
- Renamed main function to `compare_distributions`.

## Version 0.1.0
- Added main function, `variable_summaries`.

## Version 0.0.2
- Created tool to generate reports from stats and data.

## Version 0.0.1
- The initial creation of the package.


tests/test-all.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

library("testthat")
library("vardist")
test_check("vardist")


R/default_tests.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

within_tolerance <- function(tolerance, comparison = "<>", relative_diff = TRUE) {

  diff <- function(baseline, challenger) {
    if (any(is.na(c(baseline, challenger)))) {
      warning("Unable to generate a diff with NA values")
      return(NA)
    }
    if (baseline == 0 && challenger == 0) { return(0) }
    if (relative_diff) {
      (challenger - baseline) / baseline
    } else { challenger - baseline }
  }

  test_functions <- list(
    ">" = function(baseline_value, challenger_value, ...){
      diff(baseline_value, challenger_value) < tolerance
    },
    "<" = function(baseline_value, challenger_value, ...){
      -diff(baseline_value, challenger_value) < tolerance
    },
    "<>" = function(baseline_value, challenger_value, ...){
      abs(diff(baseline_value, challenger_value)) < tolerance
    }
  )

  allowed_comparisons <- names(test_functions)
  if (!comparison %in% allowed_comparisons) {
    stop(sQuote(comparison), " must be one of ",
         paste(sQuote(allowed_comparisons), collapse = ", "), ".")
  }

  test_functions[[comparison]]
}

ks_test_fn <- function(threshold) {
  function(baseline_col, challenger_col) {
    stats::ks.test(challenger_col, baseline_col)$statistic < threshold
  }
}

accept_uniques_fn <- function(threshold, tolerance) {
  function(baseline_value, challenger_value, baseline_col, challenger_col) {
    if (baseline_value > threshold && challenger_value > threshold) {
      return(TRUE)
    }
    abs(baseline_value - challenger_value) < threshold
  }
}

default_numeric_tests <- function(tolerance, ks_test_threshold) {
  list(
    mean               = within_tolerance(tolerance),
    standard_deviation = within_tolerance(tolerance),
    min                = within_tolerance(tolerance, comparison = "<"),
    first_quartile     = within_tolerance(tolerance),
    median             = within_tolerance(tolerance),
    third_quartile     = within_tolerance(tolerance),
    max                = within_tolerance(tolerance, comparison = ">"),
    "percentile_10"    = within_tolerance(tolerance),
    "percentile_90"    = within_tolerance(tolerance),
    missing_pct        = within_tolerance(tolerance, relative_diff = FALSE),
    zero_pct           = within_tolerance(tolerance, relative_diff = FALSE),
    ks_test            = ks_test_fn(ks_test_threshold)
  )
}

default_character_tests <- function(discrete_tolerance, percent_tolerance, uniques_threshold, uniques_tolerance) {
  list(
    mean_length     = within_tolerance(discrete_tolerance, relative_diff = FALSE),
    min_length      = within_tolerance(discrete_tolerance, relative_diff = FALSE),
    max_length      = within_tolerance(discrete_tolerance, relative_diff = FALSE),
    zero_length_pct = within_tolerance(percent_tolerance, relative_diff = FALSE),
    missing_pct     = within_tolerance(percent_tolerance, relative_diff = FALSE),
    uniques         = accept_uniques_fn(uniques_threshold, uniques_tolerance)
  )
}


tests/testthat/test-default_tests.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

context("default_tests")

describe("within_tolerance", {
  test_that("it can test for relative difference between numbers", {
    baseline         <- 100
    challenger_under <- 90
    challenger_over  <- 110

    reject_over       <- within_tolerance(0.05, ">")
    reject_under      <- within_tolerance(0.05, "<")
    reject_different  <- within_tolerance(0.05, "<>")

    expect_true(reject_over(baseline, challenger_under))
    expect_false(reject_over(baseline, challenger_over))

    expect_false(reject_under(baseline, challenger_under))
    expect_true(reject_under(baseline, challenger_over))

    expect_false(reject_different(baseline, challenger_under))
    expect_false(reject_different(baseline, challenger_over))

    big_baseline                <- 1000
    big_challenger_over         <- 1010
    big_challenger_further_over <- 1100

    expect_true(reject_over(big_baseline, big_challenger_over))
    expect_false(reject_over(big_baseline, big_challenger_further_over))
  })

  test_that("it can test for absolute difference between numbers", {
    small_baseline                <- 5
    small_challenger_over         <- 6
    small_challenger_further_over <- 8
    big_baseline                  <- 100
    big_challenger_over           <- 101
    big_challenger_further_over   <- 103

    less_than_two_apart <- within_tolerance(2, relative_diff = FALSE)

    expect_true(less_than_two_apart(small_baseline, small_challenger_over))
    expect_false(less_than_two_apart(small_baseline, small_challenger_further_over))

    expect_true(less_than_two_apart(big_baseline, big_challenger_over))
    expect_false(less_than_two_apart(big_baseline, big_challenger_further_over))
  })
})


R/utils.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

get_common_columns <- function(df1, df2) {
  common_columns <- intersect(names(df1), names(df2))
  if (length(common_columns) == 0) {
    stop("The data frames provided have no common column names, ",
         "cannot calculate summaries.")
  }
  common_columns
}

subset_columns <- function(colnames, dataframe) {
  current_names <- names(dataframe)
  if (length(setdiff(colnames, current_names)) > 0) {
    stop("You passed in column names that were not in the data frame. This is bad!")
  }
  count_removed_columns <- length(setdiff(current_names, colnames))
  if (count_removed_columns > 0) {
    warning(deparse(substitute(dataframe)), " has ", count_removed_columns, " columns ",
            "not listed in ", deparse(substitute(colnames)), ". Removing extra columns.")
    dataframe[colnames]
  } else { dataframe }
}

remove_non_numeric_cols <- function(dataframe) {
  numeric_columns <- names(dataframe)[vapply(dataframe, is.numeric, logical(1))]
  subset_columns(numeric_columns, dataframe)
}

remove_non_character_cols <- function(dataframe) {
  character_columns <- names(dataframe)[vapply(dataframe, is.character, logical(1))]
  subset_columns(character_columns, dataframe)
}

subset_variables <- function(varnames, dataframe) {
  current_vars <- dataframe$variables
  if (is.null(current_vars)) {
    stop("The ", sQuote("variables"), " column in ", deparse(substitute(dataframe)),
         "is NULL, either your data frame is empty or ",
         "it has no column called ", sQuote("variables"), ".")
  }
  if (length(setdiff(varnames, current_vars))) {
    stop("You passed in variable names that were not in the data frame. This is bad!")
  }
  count_removed_vars <- length(setdiff(current_vars, varnames))
  if (count_removed_vars > 0) {
    warning(deparse(substitute(dataframe)), " has ", count_removed_vars, " variables ",
            "not listed in ", deparse(substitute(colnames)), ". Removing extra variables")
    is_in_varnames <- vapply(dataframe$variables, function(x) { x %in% varnames }, logical(1))
    dataframe[is_in_varnames, ]
  } else { dataframe }

}



R/generate_report.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

#' Compare summary statistics for two datasets, given the summaries and the datasets.
#'
#' In the list \code{tests}, some elements may be named after some of the
#' summaries (which at this point are column names in \code{baseline_stats}).
#'
#' If an element in \code{tests} is named after a summary, then the element is
#' expected to be a function that takes in the baseline statistic, the challenger
#' statistic, and the respective columns. Otherwise, it is expected to take in
#' the two columns.
#'
#' @param challenger_stats data.frame. Summaries for the challenger data frame,
#'   created by \code{calculate_summaries}.
#' @param baseline_stats data.frame. Summaries for the baseline data frame,
#'   created by \code{calculate_summaries}.
#' @param challenger_df data.frame.
#' @param baseline_df data.frame.
#' @param tests list. A named list of functions that compare the challenger and
#'   the baseline. See above for details.
#' @param parallel logical. Should we use \code{mclapply} instead of \code{lapply}?
#' @param mc.cores numeric. To be passed into \code{mclapply}.
#' @return a data.frame of logical values showing if each test passed or failed
#'   for each column.
generate_report <- function(challenger_stats, baseline_stats, challenger_df,
                     baseline_df, tests, parallel = FALSE, mc.cores = parallel::detectCores()) {

  stopifnot(setequal(baseline_stats$variables, names(baseline_df)))
  stopifnot(setequal(names(baseline_df), names(challenger_df)))
  stopifnot(setequal(baseline_stats$variables, challenger_stats$variables))
  stopifnot(setequal(names(baseline_stats), names(challenger_stats)))

  report <- list()
  report$variables = baseline_stats$variables

  calculate_test_results <- function(idx) {
    map_method <- if (isTRUE(parallel)) { parallel::mcMap } else { Map }
    make_args <- function(lst) {
      if (isTRUE(parallel)) {
        append(lst, list(mc.cores = mc.cores))
      } else { lst }
    }
    if (idx %in% names(baseline_stats)) {
      baseline_results   <- baseline_stats[[idx]]
      challenger_results <- challenger_stats[[idx]]
      args <- make_args(list(tests[[idx]], baseline_results, challenger_results,
        baseline_df, challenger_df))
      unlist(do.call(map_method, args))
    } else {
      args <- make_args(list(tests[[idx]], baseline_df, challenger_df))
      unlist(do.call(map_method, args))
    }
  }
  report[names(tests)] <- `names<-`(lapply(names(tests),  calculate_test_results), names(tests))
  as.data.frame(report, stringsAsFactors = FALSE) %>% `row.names<-`(., NULL)
}


#' Compare summary statistics for two datasets, given only the summaries.
#'
#' @param challenger_stats data.frame. Summaries for the challenger data frame,
#'   created by \code{calculate_summaries}.
#' @param baseline_stats data.frame. Summaries for the baseline data frame,
#'   created by \code{calculate_summaries}.
#' @param challenger_df data.frame.
#' @param baseline_df data.frame.
#' @param tests list. A named list of functions that compare the challenger and
#'   the baseline. See above for details.
#' @return a data.frame of logical values showing if each test passed or failed
#'   for each column.
#' @export
generate_report_from_summaries <- function(challenger_stats, baseline_stats, tests) {
  common_stats     <- intersect(names(baseline_stats), names(challenger_stats))
  if (length(setdiff(names(tests), common_stats)) > 0) {
    stop("You've specified tests that don't match to columns in your statistics.")
  }
  baseline_stats   <- subset_columns(common_stats, baseline_stats)
  challenger_stats <- subset_columns(common_stats, challenger_stats)

  common_variables <- intersect(baseline_stats$variables, challenger_stats$variables)
  baseline_stats   <- subset_variables(common_variables, baseline_stats)
  challenger_stats <- subset_variables(common_variables, challenger_stats)

  report <- list()
  report$variables <- baseline_stats$variables
  report[names(tests)] <- lapply(names(tests), function(idx) {
    baseline_results   <- baseline_stats[[idx]]
    challenger_results <- challenger_stats[[idx]]
    Map(tests[[idx]], baseline_results, challenger_results) %>% unlist
  }) %>% setNames(., names(tests))

  as.data.frame(report, stringsAsFactors = FALSE) %>% `row.names<-`(., NULL)
}


R/default_summaries.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

missing_pct <- function(col) { mean(is.na(col)) }

zero_pct <- function(col) {
  vapply(col, function(x) identical(x, 0), logical(1)) %>% mean
}

quantile_fn <- function(x) {
  stopifnot(length(x) == 1)
  function(col) {
    quantile(col, x, na.rm = TRUE)
  }
}

five_number_summary <- list(
  min            = function(col) { min(col, na.rm = TRUE) },
  first_quartile = quantile_fn(0.25),
  median         = quantile_fn(0.5),
  third_quartile = quantile_fn(0.75),
  max            = function(col) { max(col, na.rm = TRUE) }
)

mean_and_sd <- list(
  mean               = function(col) { mean(col, na.rm = TRUE) },
  standard_deviation = function(col) { sd(col, na.rm = TRUE) }
)

default_numeric_summaries <- c(
  mean_and_sd,
  five_number_summary,
  list(
    "percentile_10" = quantile_fn(0.1),
    "percentile_90" = quantile_fn(0.9),
    "missing_pct"   = missing_pct,
    "zero_pct"      = zero_pct
  )
)

# if we don't pass in `keepNA = TRUE`, and `NA` will be replaced
# by a zero - but we want to distinguish between `NA` and `""`!
nchar_ <- function(x) nchar(x, keepNA = TRUE)

string_length_summaries <- list(
  mean_length     = function(col) { nchar_(col) %>% mean(., na.rm = TRUE) },
  min_length      = function(col) { nchar_(col) %>% min(., na.rm = TRUE) },
  max_length      = function(col) { nchar_(col) %>% max(., na.rm = TRUE) },
  zero_length_pct = function(col) { nchar_(col) %>% zero_pct }
)

default_character_summaries <- c(
  string_length_summaries,
  list(
    missing_pct   = missing_pct,
    count_uniques = function(col) {length(unique(col))}
  )
)





R/calculate_summaries.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

#' Calculate columnwise summaries for a data frame.
#'
#' @param dataframe data.frame. The data you want to summarize.
#' @param summaries list. A list of summary functions. Each function must take
#'   as an input one numeric vector, and output a numeric vector of length 1.
#' @param parallel logical. Should we use \code{mclapply} instead of \code{lapply}?
#' @param mc.cores numeric. To be passed into \code{mclapply}.
#' @return data.frame, with a column called \code{variables} with the column names
#'   from the data frame, and a column for each summary function.
#' @export
calculate_numeric_summaries <- function(dataframe, summaries = default_numeric_summaries, parallel = FALSE, mc.cores = parallel::detectCores()) {
  dataframe <- remove_non_numeric_cols(dataframe)
  calculate_summaries(dataframe, summaries, parallel, mc.cores)
}

#' Calculate columnwise summaries for a data frame.
#' @inheritParams calculate_numeric_summaries
#' @export
calculate_character_summaries <- function(dataframe, summaries = default_character_summaries, parallel = FALSE, mc.cores = parallel::detectCores()) {
  dataframe <- remove_non_character_cols(dataframe)
  calculate_summaries(dataframe, summaries, parallel, mc.cores)
}

calculate_summaries <- function(dataframe, summaries, parallel = FALSE, mc.cores = parallel::detectCores()) {
  output <- list()
  output$variables <- names(dataframe)
  if (isTRUE(parallel)) {
    output[names(summaries)] <- lapply(summaries, function(fn) {
      unlist(parallel::mclapply(dataframe, fn, mc.cores = mc.cores))
    })
  } else {
    output[names(summaries)] <- lapply(summaries, function(fn) {
      vapply(dataframe, fn, numeric(1))
    })
  }
  as.data.frame(output, stringsAsFactors = FALSE) %>% `row.names<-`(., NULL)
}


R/compare_distributions.R
# Copyright 2017 Avant
# This file is licensed under the MIT license. For a full copy of the license, see: 
# https://github.com/avantoss/open-source/blob/master/LICENSE_MIT

#' Compare the distribution of common fields across two data frames.
#'
#' Typically you have a data set whose integrity is unknown, and you want to compare
#' it to a data set whose reliability has already been established by other means.
#' With this function, you can compare the uncertain data set (the "challenger") to
#' the certain one (the "baseline") and see if they have similar enough distributions.
#'
#' @param challenger data.frame.
#' @param baseline data.frame.
#' @param summaries list. A named list of summary functions. Each function must take
#'   as an input one numeric vector, and output a numeric vector of length 1.
#' @param tests list. A named list of functions that return \code{TRUE} or \code{FALSE}, and take in
#'   columns or summary statistics fom the \code{challenger} and the \code{baseline}.
#' @param parallel logical. Should we use \code{mclapply} instead of \code{lapply}?
#' @param mc.cores numeric. To be passed into \code{mclapply}.
#' @return a list with three data frames - the columnwise summaries for the \code{challenger},
#'   the columnwise summaries for the \code{baseline}, and a report with the results of the
#'   \code{tests}.
#' @seealso calculate_summaries, generate_report
#' @export
compare_numeric_distributions <- function(challenger, baseline,
                           summaries = default_numeric_summaries,
                           tests     = default_numeric_tests(
                             tolerance         = getOption("vardist.numeric_summary_tolerance", 0.1),
                             ks_test_threshold = getOption("vardist.ks_test_threshold", 0.5)),
                           parallel = FALSE,
                           mc.cores = parallel::detectCores()) {
  challenger <- remove_non_numeric_cols(challenger)
  baseline   <- remove_non_numeric_cols(baseline)
  calculate_summaries_and_report(challenger, baseline, summaries, tests, parallel, mc.cores)
}

#' Compare the distribution of common fields across two data frames.
#' @inheritParams compare_numeric_distributions
#' @seealso compare_numeric_distributions
#' @export
compare_character_distributions <- function(challenger, baseline,
                           summaries = default_character_summaries,
                           tests     = default_character_tests(
                             discrete_tolerance = getOption("vardist.character_discrete_tolerance", 2),
                             percent_tolerance  = getOption("vardist.character_percent_tolerance", 0.1),
                             uniques_threshold  = getOption("vardist.character_uniques_threshold", 100),
                             uniques_tolerance  = getOption("vardist.character_uniques_tolerance", 10)),
                           parallel = FALSE,
                           mc.cores = parallel::detectCores()) {
  challenger <- remove_non_character_cols(challenger)
  baseline   <- remove_non_character_cols(baseline)
  calculate_summaries_and_report(challenger, baseline, summaries, tests, parallel, mc.cores)
}

calculate_summaries_and_report <- function(challenger, baseline, summaries, tests, parallel, mc.cores) {
  common_columns <- get_common_columns(challenger, baseline)
  challenger     <- subset_columns(common_columns, challenger)
  baseline       <- subset_columns(common_columns, baseline)

  challenger_stats <- calculate_summaries(challenger, summaries, parallel, mc.cores)
  baseline_stats   <- calculate_summaries(baseline, summaries, parallel, mc.cores)
  report           <- generate_report(challenger_stats, baseline_stats,
                        challenger, baseline, tests, parallel, mc.cores)

  list(
    challenger_stats = challenger_stats,
    baseline_stats   = baseline_stats,
    report           = report
  )
}
